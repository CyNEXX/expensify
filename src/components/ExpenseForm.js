import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';


const now = moment();

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            text: props.expense ? props.expense.text : '',
            amount: props.expense ? (props.expense.amount).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            editForm: props.editForm
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;

        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }

    onDateChange = (createdAt) => {
        if (createdAt) { this.setState(() => ({ createdAt })); }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    }

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };

    render() {
        return (
            <form onSubmit={this.onSubmit}>
                {this.state.error && <p className='font-italic text-danger'>{this.state.error}</p>}
                <div className='d-flex flex-column form'>
                    <div className='d-flex flex-wrap col-12'>
                        <input
                            className='basic-input w-100 px-3'
                            type='text'
                            placeholder='Decription'
                            autoFocus
                            value={this.state.description}
                            onChange={this.onDescriptionChange}>
                        </input>
                    </div>
                    <div className='d-flex flex-sm-column flex-xl-row justify-content-between col-12'>
                        <div /* className='col-sm-12 col-xl-6' */ className='flex-fill mr-xl-2 mb-sm-2 mb-xl-0'>
                            <input
                                className='basic-input w-100 px-3'
                                type='text'
                                value={this.state.amount}
                                placeholder='Amount'
                                onChange={this.onAmountChange}>
                            </input>
                        </div>
                        <div className='date-container' className='flex-fill ml-xl-2 mt-sm-2 mt-xl-0'/* col-sm-12 col-xl-6' */>
                            <SingleDatePicker
                                date={this.state.createdAt} // momentPropTypes.momentObj or null
                                onDateChange={this.onDateChange} // PropTypes.func.isRequired
                                focused={this.state.calendarFocused} // PropTypes.bool
                                onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
                                numberOfMonths={1}
                                isOutsideRange={(day) => false}
                            // id="your_unique_id" // PropTypes.string.isRequired,
                            />
                        </div>
                    </div>
                    <div className='d-flex col-12'>
                        <textarea
                            className='basic-textarea w-100 px-3'
                            type='text'
                            placeholder='Add a description for this expense (optional)'
                            value={this.state.note}
                            onChange={this.onNoteChange}>
                        </textarea>
                    </div>
                    <div className='mb-0 px-3 col-xl-3 col-sm-12'>
                        <button className='btn btn-primary w-100'>
                            {this.state.editForm ? 'Save Expense' : 'Add Expense'}
                        </button>
                    </div>
                </div>

            </form>
        )
    }
}