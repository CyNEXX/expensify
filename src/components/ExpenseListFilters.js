import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';



export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => { this.setState(() => ({ calendarFocused })); }

    ontTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        }
        else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    }



    render() {
        return (
            <div className='col-12 d-flex content-container my-4 flex-wrap'>
                <div className='col-sm-12 col-xl-6 px-0'>
                    <input className='w-100 pl-3 basic-input'
                        placeholder='Search expense...'
                        type='text'
                        value={this.props.filters.text}
                        onChange={this.ontTextChange}>
                    </input>
                </div>
                <div className='col-sm-4 col-xl-2 px-xl-3 mt-sm-3 mt-xl-0 pl-sm-0'>
                    <select className='w-100 pl-3 basic-input'
                        value={this.props.filters.sortBy}
                        onChange={this.onSortChange}>
                        <option value="date">Date</option>
                        <option value="amount">Amount</option>
                    </select>
                </div>
                <div className='col-xl-4 col-sm-5 px-0 mt-sm-3 mt-xl-0 datepicker'>
                    <DateRangePicker
                        startDate={this.props.filters.startDate}
                        startDateId='startD'
                        endDate={this.props.filters.endDate}
                        endDateId='endD'
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        showClearDates={true}
                        numberOfMonths={1}
                        isOutsideRange={(day) => {
                            return false;
                        }}
                    ></DateRangePicker>
                </div>
            </div>
        );
    }
}




const mapStatetoProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStatetoProps, mapDispatchToProps)(ExpenseListFilters);