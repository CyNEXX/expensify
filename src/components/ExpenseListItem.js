import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ amount, createdAt, count, description, id }) => (
    <Link className='col-12 fieldgroup flex-xl-row flex-sm-column border border-top-0' to={`/edit/${id}`}>
        <div className='py-3 d-flex flex-column pr-4'>
            <h3 className='text-dark font-weight-bold text-break'>{description}</h3>
            <small className='text-muted'>
                {moment(createdAt).format('DD MMMM YYYY')}
            </small>
        </div>
        <div className='my-auto h5 font-weight-bold pb-3'>
            {numeral(amount).format('$0,0.00')}
        </div>

    </Link>
);

export default ExpenseListItem;

