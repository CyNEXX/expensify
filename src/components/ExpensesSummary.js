import React from 'react';
import numeral from 'numeral';
import selectAllExpenses from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export const ExpensesSummary = ({ count, total }) => {
    return (
        <div className='bg-light py-4'>
            <div className='content-container py-2 w-100'>
                <div className='d-flex flex-sm-row flex-xl-column justify-content-between h-100'>
                    <div className='col-auto py-xl-0 py-sm-1 h-100'>
                        {
                            count == 0 ?
                                <h2 className='font-weight-light p-0 h-100'>No expenses within range</h2> :
                                <div>
                                    <div className='d-sm-none d-xl-block'>
                                        <h2 className='font-weight-light p-0 h-100'>Viewing <span className='font-weight-bold'>{count}</span> {count !== 1 ? 'expenses' : 'expense'} totalling <span className='font-weight-bold'>{numeral(total).format('$0,0.00')}</span></h2>
                                    </div>
                                    <div className='d-sm-flex d-xl-none flex-column'>
                                        <div>
                                            <span className='font-weight-bold h2'>{count}</span><span className='h2'> {count !== 1 ? 'expenses' : 'expense'} </span>
                                        </div>
                                        <h1 className='font-weight-bold pt-2'>{numeral(total).format('$0,0.00')}</h1>
                                    </div>
                                </div>
                        }
                    </div>
                    <div className='col-3 d-flex'>
                        <Link className='btn btn-primary mt-sm-0 mt-xl-3 bigbutton w-100 h-100 px-0 d-flex justify-content-center align-content-center' to='/create'><h2 className='m-auto'>Add Expense</h2></Link>
                    </div>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => {
    const expenses = selectExpenses(state.expenses, state.filters);
    return {
        count: expenses.length,
        total: selectAllExpenses(expenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);

