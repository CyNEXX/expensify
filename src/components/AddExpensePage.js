import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    }
    render() {
        return (<div>
            <div className='px-3 bg-light py-4'>
                <div className='content-container py-2'>
                    <h1>Add Expense</h1>
                </div>
            </div>
            <div className='content-container my-4'>
                <ExpenseForm
                    /*                     editForm={false} */
                    onSubmit={this.onSubmit}
                />
            </div>
        </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);