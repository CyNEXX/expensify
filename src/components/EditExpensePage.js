import React from 'react';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {

    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <div className='px-3 bg-light py-4'>
                    <div className='content-container py-2'>
                        <h1>Edit Expense</h1>

                    </div>
                </div>
                <div className='content-container my-4'>
                    <div className='d-flex flex-column'>
                        <div>
                            <ExpenseForm
                                expense={this.props.expense}
                                editForm={true}
                                onSubmit={this.onSubmit}
                            />
                        </div>
                        <div className='col-sm-12 col-xl-3'>
                            <button className='btn btn-darkred w-100' onClick={this.onRemove}>Remove Expense</button>
                        </div>
                    </div>
                </div>

            </div>
        )
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchtoProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => { dispatch(startRemoveExpense(data)) }
});

export default connect(mapStateToProps, mapDispatchtoProps)(EditExpensePage);