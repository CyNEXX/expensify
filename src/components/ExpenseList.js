import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div className='fieldgroup col-12 flex-column'>
        <div className='fieldgroup col-12 bg-dark basic-input border text-light py-3'>
            <div className='pl-1 my-auto d-none d-sm-block d-xl-none'>Expenses</div>
            <div className='pl-1 my-auto d-sm-none d-xl-block'>Expense</div>
            <div className='pr-1 my-auto d-sm-none d-xl-block'>Amount</div>
        </div>
        <div className='fieldarray mb-5'>
            {
                props.expenses.length === 0 ? (
                    <div className='d-flex py-3 font-italic text-muted justify-content-center'>
                        <h5>No expense</h5>
                    </div>) : (
                    props.expenses.map((expense, i) => {
                        return <ExpenseListItem
                            key={expense.id}
                            count={i + 1}
                            {...expense} />
                    })
                )
            }
        </div>
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList);