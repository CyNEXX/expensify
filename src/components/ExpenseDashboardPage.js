import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
    <div className='d-flex justify-content-between flex-column'>
        <ExpensesSummary />
        <div className='fieldgroup content-container flex-column w-100'>
            <ExpenseListFilters />
            <ExpenseList />
        </div>
    </div>
);

export default ExpenseDashboardPage;