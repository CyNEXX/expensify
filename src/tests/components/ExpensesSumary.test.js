import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('Should render ExpensesSummary correctly with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary count={1} total={255} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should render ExpensesSummary correctly with mode expenses', () => {
    const wrapper = shallow(<ExpensesSummary count={8} total={25546848} />);
    expect(wrapper).toMatchSnapshot();
});