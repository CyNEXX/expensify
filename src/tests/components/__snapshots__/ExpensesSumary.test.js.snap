// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpensesSummary correctly with 1 expense 1`] = `
<div
  className="bg-light py-4"
>
  <div
    className="content-container py-2 w-100"
  >
    <div
      className="d-flex flex-sm-row flex-xl-column justify-content-between h-100"
    >
      <div
        className="col-auto py-xl-0 py-sm-1 h-100"
      >
        <div>
          <div
            className="d-sm-none d-xl-block"
          >
            <h2
              className="font-weight-light p-0 h-100"
            >
              Viewing 
              <span
                className="font-weight-bold"
              >
                1
              </span>
               
              expense
               totalling 
              <span
                className="font-weight-bold"
              >
                $255.00
              </span>
            </h2>
          </div>
          <div
            className="d-sm-flex d-xl-none flex-column"
          >
            <div>
              <span
                className="font-weight-bold h2"
              >
                1
              </span>
              <span
                className="h2"
              >
                 
                expense
                 
              </span>
            </div>
            <h1
              className="font-weight-bold pt-2"
            >
              $255.00
            </h1>
          </div>
        </div>
      </div>
      <div
        className="col-3 d-flex"
      >
        <Link
          className="btn btn-primary mt-sm-0 mt-xl-3 bigbutton w-100 h-100 px-0 d-flex justify-content-center align-content-center"
          to="/create"
        >
          <h2
            className="m-auto"
          >
            Add Expense
          </h2>
        </Link>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render ExpensesSummary correctly with mode expenses 1`] = `
<div
  className="bg-light py-4"
>
  <div
    className="content-container py-2 w-100"
  >
    <div
      className="d-flex flex-sm-row flex-xl-column justify-content-between h-100"
    >
      <div
        className="col-auto py-xl-0 py-sm-1 h-100"
      >
        <div>
          <div
            className="d-sm-none d-xl-block"
          >
            <h2
              className="font-weight-light p-0 h-100"
            >
              Viewing 
              <span
                className="font-weight-bold"
              >
                8
              </span>
               
              expenses
               totalling 
              <span
                className="font-weight-bold"
              >
                $25,546,848.00
              </span>
            </h2>
          </div>
          <div
            className="d-sm-flex d-xl-none flex-column"
          >
            <div>
              <span
                className="font-weight-bold h2"
              >
                8
              </span>
              <span
                className="h2"
              >
                 
                expenses
                 
              </span>
            </div>
            <h1
              className="font-weight-bold pt-2"
            >
              $25,546,848.00
            </h1>
          </div>
        </div>
      </div>
      <div
        className="col-3 d-flex"
      >
        <Link
          className="btn btn-primary mt-sm-0 mt-xl-3 bigbutton w-100 h-100 px-0 d-flex justify-content-center align-content-center"
          to="/create"
        >
          <h2
            className="m-auto"
          >
            Add Expense
          </h2>
        </Link>
      </div>
    </div>
  </div>
</div>
`;
